New blog post for pre/post command plugin hooks in conda `23.7.0` (#156)
conda-incubator/conda-dot-org: blog/2023-07-31-new-pre-post-command-plugin-hooks.mdx

---
title: "New pre and post command plugin hooks"

> slug: "2023-07-05-latest-conda-release-includes-new-plugin-hooks"

Reviewer: ```suggestion
slug: "2023-07-31-latest-conda-release-includes-new-plugin-hooks"
```

[conda-plugin-docs]: https://docs.conda.io/projects/conda/en/stable/dev-guide/plugins/index.html
[matrix]: https://matrix.to/#/#conda:matrix.org

import Image from '@theme/IdealImage';
import banner from '@site/static/img/blog/2023-07-31-new-pre-post-command-plugin-hooks/pre-post-command-image.jpg';

<Image img={banner} alt="Banner image for blog post; New pre and post command plugin hooks - picture of a plugin outlet"/>

<p>
    <i>
        <small>
            Image credit: <a href="https://unsplash.com/photos/7d8pxcMVl7A">Kier in Sight Archives on Unsplash</a>
        </small>
    </i>
</p>

With the latest conda release (`23.7.0` at the time this post was written), the ability to define
two new plugin hooks was introduced: "pre command" and "post command". These two new plugin hooks give
plugin authors the ability execute code before and after conda commands are run. In this blog post,
we provide more details on how and why you may use these to extend the default behavior of conda.

To see a fully functioning example of how to use these new hooks, please check out the [conda-guard
project][conda-guard].

<!-- truncate -->

> To explain how these plugin hooks can be used, we walk through two examples.

Reviewer: ```suggestion
To explain how these plugin hooks can be used, let's walk through two examples.
```

Reviewer: Take or leave this suggestion, it's mostly a "conversational style" change

Author: I think I'm going to change it to `... we will cover...`

## Conda guard and the "pre command" hook

> The project linked to above is called conda-guard, and it extends conda by adding functionality to "guard"

Reviewer: I know it's linked above, but having it immediately accessible while reading it is better than making the reader scroll up and search for it

## Simple analytics with "post command"

The "post command" hook functions in exactly the same way as the "pre command" except that it gets
called once a conda command has successfully finished running. It is important to note that if
the command exits early for any reason, this plugin hook will not be invoked.

To illustrate how this is used, we will create a simple plugin that counts how many times we have
run a particular conda command. We can use this later to analyze our usage of conda 🤓 📊

Here is a snippet of how that plugin might look:

```python
from conda.cli.conda_argparse import BUILTIN_COMMANDS
from conda.plugins import hookimpl, CondaPostCommand

ENV_COMMANDS = {"env_config", "env_create", "env_export", "env_list", "env_remove", "env_update"}

def conda_stats_post_commands_action(command: str):
    """Counts how many times we have run a particular conda command"""
    database = get_database()
    database.add_command(command)

@hookimpl
def conda_post_commands():
    yield CondaPostCommand(
        name=f"conda_stats_post_command",
        action=conda_stats_post_commands_action,
        run_for=BUILTIN_COMMANDS.union(ENV_COMMANDS),
    )
```

> The example above registers our "post command" hook by defining a function called
> `conda_post_commands`. Very similarly to the "pre command" hook, this returns a 
> `CondaPostCommand` object which defines "name", "action" and "run_for". The "action"
> function counts the usage of the command each time it is run and the "run_for" property
> is configured to run for every builtin conda command, including the `conda env *` 
> commands.

Reviewer: ```suggestion
The example above registers our "post command" hook by defining a function called
`conda_post_commands`. Very similarly to the "pre command" hook, this returns a 
`CondaPostCommand` object which defines `name`, `action` and `run_for`. The `action`
function counts the usage of the command each time it is run and the `run_for` property
is configured to run for every built-in conda command, including the `conda env *` 
commands.
```

The implementation of the database has been purposely left out of the example as it

> falls out of scope of this blog post (implement it yourself for fun! 😅).

Reviewer: Can we make this a "Note" section?

## Wrapping up

If you would like to get started with using these new plugins, please check out the 
[conda-guard][conda-guard] project. This project can also be used as a starting template
for your plugins.

> For even more information about all plugin hooks currently available in conda, you
> head over to the [documentation][conda-plugin-docs].

Reviewer: ```suggestion
For even more information about all plugin hooks currently available in conda,
head over to the [relevant documentation page][conda-plugin-docs].
```

---
title: "New pre and post command plugin hooks"
slug: "2023-07-31-latest-conda-release-includes-new-plugin-hooks"
description: |
  The latest release of conda introduces two new plugin hooks: pre and post command. These hooks allow plugin
  authors to execute code before and after any conda command. Learn more about how and why you would use
  these in this post.
authors: [travishathaway]
tags: [conda, plugins]
image: img/blog/2023-07-31-new-pre-post-command-plugin-hooks/pre-post-command-image.jpg
---

[conda-issue]: https://github.com/conda/conda/issues/12138

> [conda-guard]: https://github.com/travishathaway/conda-guard

Reviewer 2: Hmm, hmm, given the potential use of the name elsewhere, could we not mention that?

Author: I understand the concern. I really want to link to an example project though because I think it will enhance the experience.

Maybe a new name is warranted? What about `conda schutz`? 😂 🇩🇪. It's German so there should be any naming conflicts. Plus, it's short and easy to type.

I'll keep trying to find other names for the project...

Author: I'm going to go with "schutz" 😄.

Author: Okay, I went with "conda-protect" instead.

Reviewer 2: OMG, only see this now! "schutz" has some unfortunate historic connotations, let's try to prevent that :D

plugin authors the ability to execute code before and after conda commands are run. In this blog post,
we provide more details on how and why you may use these to extend the default behavior of conda.

<!-- truncate -->

:::info

To check out a fully functional example, check out the [conda-guard][conda-guard] project
to see how these new plugin hooks are used in practice.
:::

---

To explain how these plugin hooks can be used, we will cover two examples.

## Conda guard and the "pre command" hook

The project linked to above is called [`conda-guard`](https://github.com/travishathaway/conda-guard), and it extends conda by adding functionality to "guard"
a conda environment from being changed. Sometimes users may want to do this to protect themselves from
[unintentionally modifying their base environments][conda-issue].

Using the "pre command" plugin hook, we can first specify which commands we want it to run for and then
run a function that executes before the command gets executed. By doing this, we can figure out whether
the environment is guarded or not and exit early if so. Below is how such a plugin hook would be defined:

```python

> from conda.plugins import hookimpl, CondaPreCommand

Reviewer 2: ```suggestion
from conda import plugins
```

def conda_guard_pre_commands_action(command: str) -> None:
    """Checks to see if the current environment is guarded"""
    current_environment = get_current_environment()
    
    if is_guarded(current_environment):
        raise CondaError(
            f"Current environment is guarded. Run `conda guard {environment}` to "
            "remove guard."
        )

> @hookimpl
> def conda_pre_commands():
>     yield CondaPreCommand(
>         name=f"conda_guard_pre_command",
>         action=conda_guard_pre_commands_action,
>         run_for={"install", "remove", "update", "env_update", "env_remove"},
>     )
> ```

Reviewer 2: ```suggestion
@plugins.hookimpl
def conda_pre_commands():
    yield plugins.CondaPreCommand(
        name=f"conda_guard_pre_command",
        action=conda_guard_pre_commands_action,
        run_for={"install", "remove", "update", "env_update", "env_remove"},
    )
```

I think this makes it a little more readable since `hookimpl` is rather obtuse

In the example above, we first register our plugin hook by defining a function called
`conda_pre_commands` and then decorating it with the `hookimpl` function provided by conda.
Inside the hook function itself we return a `CondaPreCommand` object, which defines the following:

- **name**: how this plugin hook will be internally reference by conda
- **action**: a callable object that will be invoked before the commands defined in `run_for` are
  executed
- **run_for**: a `set` of commands that this plugin hook should be used with

Going back to our example, if we wanted to create a plugin that would prevent us from mistakenly
modifying "guarded" environments, it would make sense to define `run_for` as all the conda commands
that could potentially modify an environment (e.g. `install`, `remove`, etc.).

To actually prevent us from modifying our guarded environments, the `action` function
checks if the current environment is guarded and then throws a `CondaError` if it is to make
sure that the program exits early with a meaningful error message for the user.

:::info

For the full implementation, please see the [conda-guard][conda-guard] project.

:::

> ## Simple analytics with "post command"

Reviewer 2: ```suggestion
## Simple command counter with the "post command" hook
```

Let's defuse this a little, analytics is a big word and we don't want to scare people away

The "post command" hook functions in exactly the same way as the "pre command" except that it gets
called once a conda command has successfully finished running. It is important to note that if
the command exits early for any reason, this plugin hook will not be invoked.

To illustrate how this is used, we will create a simple plugin that counts how many times we have
run a particular conda command. We can use this later to analyze our usage of conda 🤓 📊

Here is a snippet of how that plugin might look:

```python
from conda.cli.conda_argparse import BUILTIN_COMMANDS

> from conda.plugins import hookimpl, CondaPostCommand

Reviewer 2: ```suggestion
from conda import plugins
```

ENV_COMMANDS = {
    "env_config", "env_create", 
    "env_export", "env_list",
    "env_remove", "env_update"
}

def conda_stats_post_commands_action(command: str):
    """Counts how many times we have run a particular conda command"""
    database = get_database()
    database.add_command(command)

> @hookimpl
> def conda_post_commands():
>     yield CondaPostCommand(
>         name=f"conda_stats_post_command",
>         action=conda_stats_post_commands_action,
>         run_for=BUILTIN_COMMANDS.union(ENV_COMMANDS),
>     )
> ```

Reviewer 2: ```suggestion
@plugins.hookimpl
def conda_post_commands():
    yield plugins.CondaPostCommand(
        name=f"conda_stats_post_command",
        action=conda_stats_post_commands_action,
        run_for=BUILTIN_COMMANDS.union(ENV_COMMANDS),
    )
```

I think this makes it a little more readable since `hookimpl` is rather obtuse

slug: "2023-07-31-latest-conda-release-includes-new-plugin-hooks"
description: |
  The latest release of conda introduces two new plugin hooks: pre and post command. These hooks allow plugin
  authors to execute code before and after any conda command. Learn more about how and why you would use
  these in this post.
authors: [travishathaway]
tags: [conda, plugins]
image: img/blog/2023-07-31-new-pre-post-command-plugin-hooks/pre-post-command-image.jpg
---

[conda-issue]: https://github.com/conda/conda/issues/12138
[conda-protect]: https://github.com/travishathaway/conda-protect
[conda-plugin-docs]: https://docs.conda.io/projects/conda/en/stable/dev-guide/plugins/index.html
[matrix]: https://matrix.to/#/#conda:matrix.org

import Image from '@theme/IdealImage';
import banner from '@site/static/img/blog/2023-07-31-new-pre-post-command-plugin-hooks/pre-post-command-image.jpg';

<Image img={banner} alt="Banner image for blog post; New pre and post command plugin hooks - picture of a plugin outlet"/>

<p>
    <i>
        <small>
            Image credit: <a href="https://unsplash.com/photos/7d8pxcMVl7A">Kier in Sight Archives on Unsplash</a>
        </small>
    </i>
</p>

> With the latest conda release (`23.7.0` at the time this post was written), the ability to define

Reviewer 2: ```suggestion
With the latest conda release (`23.7.2` at the time this post was written), the ability to define
```
